File: frontend/src/lib/wire.ts
import {
  strictJSON,
  validateRequest,
  validateDecision,
  MAX_PLAINTEXT,
} from "./protocol.ts";
// wire mirrors backend/pkg/wire (the Go source of truth) byte-for-byte on the
// JSON wire. Two layers travel on one WebSocket:
//   - Frame: the relay-visible envelope ({_relay} | {pake} | {confirm} | {box}).
//     The relay only ever sets/reads `_relay`; it forwards the rest verbatim.
//   - App messages (Request | Decision | PushSub): plaintext that lives sealed
//     inside Frame.box. Mirror pkg/wire exactly — do not diverge. See plan §5.
//
// The key-confirmation frame is {"confirm":...}; pkg/wire.Frame carries the
// same {_relay|pake|confirm|box} fields. App plaintext is space-padded to a
// fixed block before sealing (pad / PAD_BLOCK) to hide message length.

/** RelaySignal is a relay-injected control value carried in Frame._relay. */
export type RelaySignal = "peer_joined" | "peer_left" | "undeliverable";

/** Relay signals injected by the relay (clients never send these). */
export const SignalPeerJoined: RelaySignal = "peer_joined";
export const SignalPeerLeft: RelaySignal = "peer_left";
export const SignalUndeliverable: RelaySignal = "undeliverable";

/** RelaySignals lists every valid RelaySignal (mirror wire.RelaySignals). */
export const RelaySignals: readonly RelaySignal[] = [
  SignalPeerJoined,
  SignalPeerLeft,
  SignalUndeliverable,
];

/** validRelaySignal reports whether s is a known relay signal. */
export function validRelaySignal(s: string): s is RelaySignal {
  return (
    s === SignalPeerJoined || s === SignalPeerLeft || s === SignalUndeliverable
  );
}

/**
 * Frame is the JSON envelope on the WebSocket. Exactly one app field is set
 * per frame. A relay-injected frame instead carries `_relay`.
 */
export interface Frame {
  _relay?: RelaySignal;
  pake?: string; // base64 SPAKE2 Start/Finish message
  confirm?: string; // base64 SPAKE2 key-confirmation MAC
  box?: string; // base64(nonce(24) || secretbox(plaintext))
}

/** MessageKind tags an application message inside a box. */
export type MessageKind =
  "request" | "decision" | "push_sub" | "vapid_key" | "device_key" | "ack";

export const KindRequest: MessageKind = "request";
export const KindDecision: MessageKind = "decision";
export const KindPushSub: MessageKind = "push_sub";
export const KindVAPIDKey: MessageKind = "vapid_key";
export const KindDeviceKey: MessageKind = "device_key";

/** validMessageKind reports whether k is a known app message kind. */
export function validMessageKind(k: string): k is MessageKind {
  return (
    k === KindRequest ||
    k === KindDecision ||
    k === KindPushSub ||
    k === KindVAPIDKey ||
    k === KindDeviceKey ||
    k === "ack"
  );
}

/** ResponseKind is the answer shape a request asks the human for. */
export type ResponseKind = "yesno" | "choice" | "text";

export const ResponseYesNo: ResponseKind = "yesno";
export const ResponseChoice: ResponseKind = "choice";
export const ResponseText: ResponseKind = "text";

/** validResponseKind reports whether k is a known response kind. */
export function validResponseKind(k: string): k is ResponseKind {
  return k === ResponseYesNo || k === ResponseChoice || k === ResponseText;
}

/** Category is the badge shown on a request card; free-form on the wire. */
export type Category = "cash" | "deploy" | "data" | "access" | "other";

export const Categories: readonly Category[] = [
  "cash",
  "deploy",
  "data",
  "access",
  "other",
];

/** validCategory reports whether c is a known (colored) category. */
export function validCategory(c: string): c is Category {
  return (Categories as readonly string[]).includes(c);
}

/** Response describes the answer shape requested from the human. */
export interface Response {
  kind: ResponseKind;
  options?: string[]; // choice
  placeholder?: string; // text
  max_len?: number; // text
}

/** Request is an approval request sent agent -> phone, sealed inside a box. */
export interface Request {
  protocol?: number;
  room?: string;
  deadline_ms?: number;
  sig?: string;
  kind: MessageKind; // always KindRequest
  id: string;
  title: string;
  category?: Category | string;
  summary: string;
  agent?: string;
  response: Response;
  expires_in_s?: number;
}

/** Result is the human's answer; exactly one field is set per Decision. */
export interface Result {
  approved?: boolean; // yesno
  choice?: string; // choice
  text?: string; // text
}

/** Decision is the human's answer sent phone -> agent, sealed inside a box. */
export interface Decision {
  protocol?: number;
  room?: string;
  request_hash?: string;
  response_kind?: ResponseKind;
  kind: MessageKind; // always KindDecision
  id: string;
  result: Result;
  // Mandatory at the agent acceptance boundary. The codec can also represent
  // unsigned legacy/negative fixtures, which never establish authorization.
  sig?: string;
}

/** PushKeys are the client keys used to encrypt Web Push payloads (RFC 8291). */
export interface PushKeys {
  p256dh: string;
  auth: string;
}

/** PushSubscription is a Web Push subscription (RFC 8030/8291). */
export interface PushSubscription {
  endpoint: string;
  keys: PushKeys;
}

/** PushSub delivers the phone's PushSubscription to the agent, sealed. */
export interface PushSub {
  protocol?: number;
  room?: string;
  sig?: string;
  kind: MessageKind; // always KindPushSub
  subscription: PushSubscription;
}

/**
 * VapidKey delivers the agent's VAPID public key to the phone, sealed, so the
 * phone subscribes for Web Push with exactly the key the agent signs wake-up
 * pushes with (signer == subscribe-key). Only the PUBLIC key crosses the wire.
 */
export interface VapidKey {
  protocol?: number;
  room?: string;
  sig?: string;
  kind: MessageKind; // always KindVAPIDKey
  public_key: string;
}

/**
 * DeviceKey delivers the phone's per-device ECDSA P-256 PUBLIC key to the agent,
 * sealed. The phone signs every Decision with the matching non-extractable
 * private key (WebCrypto, IndexedDB) so a stolen session key cannot forge an
 * approval. Only the PUBLIC key (SPKI, base64) crosses the wire. Mirrors
 * wire.DeviceKey.
 */
export interface DeviceKey {
  kind: MessageKind; // always KindDeviceKey
  public_key: string;
}

/** AppMessage is any plaintext message that lives sealed inside Frame.box. */
export type AppMessage = Request | Decision | PushSub | VapidKey | DeviceKey;

/**
 * parseFrame parses one WebSocket text frame into a Frame, or null if it is
 * not valid JSON / not an object. The relay forwards opaque app frames
 * verbatim, so a Frame may carry any one of _relay / pake / confirm / box.
 */
export function parseFrame(raw: string): Frame | null {
  let v: unknown;
  try {
    v = JSON.parse(raw);
  } catch {
    return null;
  }
  if (typeof v !== "object" || v === null) return null;
  return v as Frame;
}

/** isRelayControl reports whether a parsed frame is a relay control frame. */
export function isRelayControl(f: Frame): f is Frame & { _relay: RelaySignal } {
  return typeof f._relay === "string" && validRelaySignal(f._relay);
}

/**
 * PAD_BLOCK is the fixed block size (bytes) the app plaintext is padded up to
 * before sealing, hiding the message length (notably yes/no approve vs decline)
 * from the content-blind relay. MUST match pkg/wire.padBlock on the Go side so
 * Go<->JS stays byte-compatible. Trailing ASCII spaces are JSON whitespace, so
 * JSON.parse / json.Unmarshal ignore them and decoders need no change.
 */
const PAD_BLOCK = 256;

/**
 * pad right-pads s with ASCII spaces to the next multiple of PAD_BLOCK (an
 * already-aligned input gets a full extra block, so the unpadded length is
 * never recoverable). Mirrors pkg/wire.pad.
 */
function pad(s: string): string {
  const length = new TextEncoder().encode(s).length;
  const n = PAD_BLOCK - (length % PAD_BLOCK);
  if (length + n > MAX_PLAINTEXT)
    throw new Error("wire: encoded message exceeds 16 KiB");
  return s + " ".repeat(n);
}

const MAX_VAPID_KEY_LEN = 256; // base64url uncompressed P-256 is ~88 chars
const MAX_DEVICE_KEY_LEN = 512; // base64 SPKI DER for P-256 is ~120 chars

function checkStr(
  v: unknown,
  name: string,
  max: number,
  required: boolean,
): string {
  if (v === undefined || v === "") {
    if (required) throw new Error(`wire: ${name} missing`);
    return "";
  }
  if (typeof v !== "string") throw new Error(`wire: ${name} must be a string`);
  if (v.length > max)
    throw new Error(`wire: ${name} too long (${v.length} > ${max})`);
  return v;
}

/** decodeRequest validates a sealed-box plaintext as a wire.Request. */
export function decodeRequest(plaintext: Uint8Array): Request {
  const msg = strictJSON(plaintext) as Request;
  validateRequest(msg);
  return msg;
}

/** decodeDecision validates a sealed-box plaintext as a wire.Decision. */
export function decodeDecision(plaintext: Uint8Array): Decision {
  const msg = strictJSON(plaintext) as Decision;
  validateDecision(msg);
  return msg;
}

/**
 * encodeDecision serializes a Decision to UTF-8 bytes for sealing, padded to a
 * fixed block so approve vs decline (and all decisions) seal to the same length
 * — the relay cannot infer the answer from ciphertext length. Mirrors
 * pkg/wire.EncodeDecision.
 */
export function encodeDecision(d: Decision): Uint8Array {
  return new TextEncoder().encode(pad(JSON.stringify(d)));
}

/** encodeRequest serializes a Request to UTF-8 bytes for sealing, padded to a
 *  fixed block to hide the request body length. Mirrors pkg/wire.EncodeRequest. */
export function encodeRequest(r: Request): Uint8Array {
  return new TextEncoder().encode(pad(JSON.stringify(r)));
}

/** encodePushSub serializes a PushSub to UTF-8 bytes for sealing, padded. */
export function encodePushSub(p: PushSub): Uint8Array {
  return new TextEncoder().encode(pad(JSON.stringify(p)));
}

/** decodeVapidKey validates a sealed-box plaintext as a wire.VapidKey. */
export function decodeVapidKey(plaintext: Uint8Array): VapidKey {
  const msg = JSON.parse(
    new TextDecoder().decode(plaintext),
  ) as Partial<VapidKey>;
  if (!validMessageKind(msg.kind ?? "") || msg.kind !== KindVAPIDKey) {
    throw new Error(`wire: not a vapid_key (kind=${String(msg.kind)})`);
  }
  checkStr(msg.public_key, "vapid_key public_key", MAX_VAPID_KEY_LEN, true);
  return msg as VapidKey;
}

/** encodeVapidKey serializes a VapidKey to UTF-8 bytes for sealing, padded. */
export function encodeVapidKey(publicKey: string): Uint8Array {
  const v: VapidKey = { kind: KindVAPIDKey, public_key: publicKey };
  return new TextEncoder().encode(pad(JSON.stringify(v)));
}

/** decodeDeviceKey validates a sealed-box plaintext as a wire.DeviceKey. */
export function decodeDeviceKey(plaintext: Uint8Array): DeviceKey {
  const msg = JSON.parse(
    new TextDecoder().decode(plaintext),
  ) as Partial<DeviceKey>;
  if (!validMessageKind(msg.kind ?? "") || msg.kind !== KindDeviceKey) {
    throw new Error(`wire: not a device_key (kind=${String(msg.kind)})`);
  }
  checkStr(msg.public_key, "device_key public_key", MAX_DEVICE_KEY_LEN, true);
  return msg as DeviceKey;
}

/** encodeDeviceKey serializes a DeviceKey to UTF-8 bytes for sealing, padded to
 *  a fixed block so its length does not leak. Mirrors encodeVapidKey. */
export function encodeDeviceKey(spkiB64: string): Uint8Array {
  const v: DeviceKey = { kind: KindDeviceKey, public_key: spkiB64 };
  return new TextEncoder().encode(pad(JSON.stringify(v)));
}

/**
 * decisionSigningMessage builds the canonical byte string the phone signs and
 * the agent verifies for one decision. It is a cross-language contract: the Go
 * twin wire.DecisionSigningMessage MUST produce byte-identical output (both are
 * pinned to the same hex in their wire tests). The message binds the room, the
 * request id, and the exact result so a signature cannot be replayed across
 * rooms, requests, or answers. Layout (no trailing separator):
 *
 *   UTF8("aah:decision:v1" 0x00 roomID 0x00 id 0x00 resultTag(result))
 *
 * resultTag picks the branch by which Result field is set, in the fixed priority
 * order approved (yesno) -> choice -> text. See ADR 0021.
 */
export function decisionSigningMessage(
  roomID: string,
  id: string,
  result: Result,
): Uint8Array<ArrayBuffer> {
  const msg = `aah:decision:v1\x00${roomID}\x00${id}\x00${resultTag(result)}`;
  return new TextEncoder().encode(msg);
}

/** resultTag renders a Result as the canonical tag used by
 *  decisionSigningMessage, matching the Go twin exactly. */
function resultTag(result: Result): string {
  if (result.approved !== undefined)
    return `yesno:${result.approved ? "1" : "0"}`;
  if (result.choice) return `choice:${result.choice}`;
  return `text:${result.text ?? ""}`;
}

/** newYesNoDecision builds a yesno Decision for request id. */
export function newYesNoDecision(id: string, approved: boolean): Decision {
  return { kind: KindDecision, id, result: { approved } };
}

/** newChoiceDecision builds a choice Decision for request id. */
export function newChoiceDecision(id: string, choice: string): Decision {
  return { kind: KindDecision, id, result: { choice } };
}

/** newTextDecision builds a text Decision for request id. */
export function newTextDecision(id: string, text: string): Decision {
  return { kind: KindDecision, id, result: { text } };
}

File: backend/pkg/wire/wire.go
// Package wire defines the JSON frames exchanged with the relay and the
// application messages that live, encrypted, inside a sealed box.
//
// Two layers travel on the same WebSocket:
//
//   - Relay frames (Frame): the relay generates `_relay` signals and
//     forwards every other field verbatim. It never decodes Box/Sealed.
//   - Application messages (Request, Decision, PushSub): plaintext that
//     lives inside Frame.Box and is only ever readable by the two paired
//     peers. The relay is content-blind. See docs/plan.md section 5.
package wire

import (
	"bytes"
	"encoding/json"
)

// padBlock is the fixed block size (bytes) the app plaintext is padded up to
// before sealing. A yes/no decision otherwise leaks via ciphertext length
// (approve vs decline differ by a few bytes); padding to a fixed multiple
// hides the distinction. Padding is trailing ASCII spaces, which both
// encoding/json and JSON.parse ignore, so decoders need no change. The JS
// side (frontend/src/lib/wire.ts pad) MUST use this same constant.
const padBlock = 256

// pad right-pads b with ASCII spaces (0x20) up to the next multiple of
// padBlock. JSON parsers ignore trailing whitespace, so a padded plaintext
// round-trips identically. An already block-aligned input is padded by a
// full block so the unpadded length is never recoverable from the total.
func pad(b []byte) []byte {
	n := padBlock - (len(b) % padBlock)
	return append(b, bytes.Repeat([]byte{' '}, n)...)
}

// EncodeRequest JSON-encodes r and pads the plaintext to a fixed block so the
// request body length does not leak through the sealed box. Seal the result.
func EncodeRequest(r Request) ([]byte, error) {
	raw, err := json.Marshal(r)
	if err != nil {
		return nil, err
	}
	return pad(raw), nil
}

// EncodeDecision JSON-encodes d and pads the plaintext to a fixed block so
// approve vs decline (and all decisions) seal to the same length, hiding the
// answer from the relay via ciphertext-length analysis. Seal the result.
func EncodeDecision(d Decision) ([]byte, error) {
	raw, err := json.Marshal(d)
	if err != nil {
		return nil, err
	}
	return pad(raw), nil
}

// EncodeVAPIDKey JSON-encodes the agent's VAPID public key and pads the
// plaintext to a fixed block, matching the other sealed encoders. Only the
// public key crosses the wire; the private key never leaves the agent. Seal
// the result.
func EncodeVAPIDKey(pub string) ([]byte, error) {
	raw, err := json.Marshal(VAPIDKey{Kind: KindVAPIDKey, PublicKey: pub})
	if err != nil {
		return nil, err
	}
	return pad(raw), nil
}

// RelaySignal is a relay-injected control value carried in Frame.Relay.
// The relay is the only party that may set it; clients never send it.
type RelaySignal string

// Relay signals injected by the relay (see docs/plan.md section 5).
const (
	// SignalPeerJoined means the room's other peer connected.
	SignalPeerJoined RelaySignal = "peer_joined"
	// SignalPeerLeft means the room's other peer disconnected.
	SignalPeerLeft RelaySignal = "peer_left"
	// SignalUndeliverable means a frame could not be delivered because the
	// peer is absent; the sender must retry. See docs/plan.md section 8.
	SignalUndeliverable RelaySignal = "undeliverable"
)

// ValidRelaySignal reports whether s is a known relay signal.
func ValidRelaySignal(s RelaySignal) bool {
	switch s {
	case SignalPeerJoined, SignalPeerLeft, SignalUndeliverable:
		return true
	default:
		return false
	}
}

// Frame is the JSON envelope on the WebSocket. Exactly one field is set
// per frame. The relay reads only Relay; it forwards Pake, Confirm, and Box
// verbatim without inspecting their contents.
type Frame struct {
	// Relay is a relay-injected control signal; clients never set it.
	Relay RelaySignal `json:"_relay,omitempty"`
	// Pake carries a base64 SPAKE2 Start/Finish message during pairing.
	Pake string `json:"pake,omitempty"`
	// Confirm carries the base64 SPAKE2 key-confirmation MAC during pairing.
	Confirm string `json:"confirm,omitempty"`
	// Box carries base64(nonce || ciphertext) for all post-pairing traffic.
	Box string `json:"box,omitempty"`
}

// MessageKind tags an application message inside a Box.
type MessageKind string

// Application message kinds (see docs/plan.md section 5).
const (
	// KindRequest is an approval request from agent to phone.
	KindRequest MessageKind = "request"
	// KindDecision is the human's answer from phone to agent.
	KindDecision MessageKind = "decision"
	// KindPushSub delivers the phone's sealed push subscription to the agent.
	KindPushSub MessageKind = "push_sub"
	// KindVAPIDKey delivers the agent's VAPID public key to the phone so it
	// subscribes with exactly the key the agent signs wake-up pushes with.
	KindVAPIDKey MessageKind = "vapid_key"
)

// ValidMessageKind reports whether k is a known application message kind.
func ValidMessageKind(k MessageKind) bool {
	switch k {
	case KindRequest, KindDecision, KindPushSub, KindVAPIDKey:
		return true
	default:
		return false
	}
}

// ResponseKind is the answer shape a request asks the human for.
type ResponseKind string

// Response kinds the phone can render (see docs/plan.md section 5).
const (
	// ResponseYesNo is a swipe approve/decline card.
	ResponseYesNo ResponseKind = "yesno"
	// ResponseChoice is a set of tappable options.
	ResponseChoice ResponseKind = "choice"
	// ResponseText is a short free-text input.
	ResponseText ResponseKind = "text"
)

// ValidResponseKind reports whether k is a known response kind.
func ValidResponseKind(k ResponseKind) bool {
	switch k {
	case ResponseYesNo, ResponseChoice, ResponseText:
		return true
	default:
		return false
	}
}

// Category is the badge shown on a request card. It is free-form on the
// wire; these are the known values that get a dedicated color.
type Category string

// Known request categories (see docs/plan.md section 5).
const (
	// CategoryCash flags money-moving requests.
	CategoryCash Category = "cash"
	// CategoryDeploy flags deployment/release requests.
	CategoryDeploy Category = "deploy"
	// CategoryData flags data-mutating requests.
	CategoryData Category = "data"
	// CategoryAccess flags access/permission requests.
	CategoryAccess Category = "access"
	// CategoryOther is the catch-all badge.
	CategoryOther Category = "other"
)

// ValidCategory reports whether c is a known category. Categories are
// free-form on the wire; an unknown value is still rendered (as "other").
func ValidCategory(c Category) bool {
	switch c {
	case CategoryCash, CategoryDeploy, CategoryData, CategoryAccess, CategoryOther:
		return true
	default:
		return false
	}
}

// Response describes the answer shape requested from the human.
type Response struct {
	// Kind selects the UI: yesno, choice, or text.
	Kind ResponseKind `json:"kind"`
	// Options are the choices for ResponseChoice.
	Options []string `json:"options,omitempty"`
	// Placeholder hints the input for ResponseText.
	Placeholder string `json:"placeholder,omitempty"`
	// MaxLen caps the input length for ResponseText.
	MaxLen int `json:"max_len,omitempty"`
}

// Request is an approval request sent agent -> phone, sealed inside a Box.
type Request struct {
	// Kind is always KindRequest.
	Kind MessageKind `json:"kind"`
	// ID uniquely identifies the request; the phone de-dupes by it.
	ID string `json:"id"`
	// Title is the card's window title.
	Title string `json:"title"`
	// Category is the badge color (see Category).
	Category Category `json:"category,omitempty"`
	// Summary is the human-readable question body.
	Summary string `json:"summary"`
	// Agent optionally names who is asking (e.g. "cursor @ workstation").
	Agent string `json:"agent,omitempty"`
	// Response is the answer shape requested.
	Response Response `json:"response"`
	// ExpiresInS is the optional countdown in seconds.
	ExpiresInS int `json:"expires_in_s,omitempty"`
}

// Result is the human's answer; exactly one field is set per Decision,
// matching the Request's Response.Kind.
type Result struct {
	// Approved is set for ResponseYesNo.
	Approved *bool `json:"approved,omitempty"`
	// Choice is set for ResponseChoice.
	Choice string `json:"choice,omitempty"`
	// Text is set for ResponseText.
	Text string `json:"text,omitempty"`
}

// Decision is the human's answer sent phone -> agent, sealed inside a Box.
type Decision struct {
	// Kind is always KindDecision.
	Kind MessageKind `json:"kind"`
	// ID echoes the Request.ID being answered.
	ID string `json:"id"`
	// Result carries the answer matching the request's response kind.
	Result Result `json:"result"`
}

// PushSubscription is a Web Push subscription (RFC 8030/8291). The phone
// sends it sealed so the relay never learns the endpoint.
type PushSubscription struct {
	// Endpoint is the push service URL.
	Endpoint string `json:"endpoint"`
	// Keys holds the p256dh and auth secrets for RFC 8291 encryption.
	Keys PushKeys `json:"keys"`
}

// PushKeys are the client keys used to encrypt Web Push payloads.
type PushKeys struct {
	// P256dh is the client's public key (base64url, uncompressed P-256).
	P256dh string `json:"p256dh"`
	// Auth is the client's auth secret (base64url, 16 bytes).
	Auth string `json:"auth"`
}

// PushSub delivers the phone's PushSubscription to the agent, sealed
// inside a Box so the relay never sees the endpoint.
type PushSub struct {
	// Kind is always KindPushSub.
	Kind MessageKind `json:"kind"`
	// Subscription is the sealed Web Push subscription.
	Subscription PushSubscription `json:"subscription"`
}

// VAPIDKey delivers the agent's VAPID public key to the phone, sealed
// inside a Box so the phone subscribes for Web Push with exactly the key
// the agent signs wake-up pushes with (signer == subscribe-key). Only the
// PUBLIC key ever crosses the wire; the private key never leaves the agent.
type VAPIDKey struct {
	// Kind is always KindVAPIDKey.
	Kind MessageKind `json:"kind"`
	// PublicKey is the agent's VAPID public key (base64url, uncompressed P-256).
	PublicKey string `json:"public_key"`
}
